### PR TITLE
Reorganize /opencl and add missing matrix_cl overloads

### LINKED ISSUE #1352: Reorganize the /opencl folder
## Description
We currently have all the OpenCL function directly in the `/opencl` folder. We have already discussed this a few time but with the coming compiler support this should be done sooner rather than later.

My proposal (open to suggestions): 
- all `matrix_cl` overloads of `/prim/mat` functions should go to `/opencl/prim`
- all future overloads of `/rev/mat` functions will go to `/opencl/rev`
- core files stay in the `/opencl` (`context, kernel, matrix_cl, ...`)
- kernels stay in `/opencl/kernels`
- other `matrix_cl` functions that are not overloads of any existing Stan Math function stay in `/opencl` (a subfolder like /helpers is another option idk).

I can tackle this one once at least @SteveBronder and @t4c1 express their opinion.

## Expected Output
More organized files.

#### Current Version:
v2.20.0

### PR DESCRIPTION
**Pull request by rok-cesnovar**
Summary

Fixes #1352 

This PR reorganizes the `/opencl` folder: 
- `matrix_cl` overloads of `/prim/mat` functions go to `/opencl/prim`
- core `OpenCL` code and functions that are not overloads of `/prim` functions stay in `/opencl`
- create a non-inplace cholesky decompose (the in-place one is moved to ::opencl)
- minor doxygen cleanup

It also creates the following signatures for matrix_cl overloads by reusing already merged OpenCL code: 
- `cholesky_decompose` in `/prim` is in the form A = cholesky_decompose(B), the one in /opencl is moved to the opencl:: namespace and still works inplace
- mdivide_left_tri_low(A,b)
- mdivide_left_tri_low(A)
- mdivide_right_tri_low(b,A)

After this PR gets merged the we will have the list of the matrix_cl overloads listed [here](https://github.com/bstatcomp/math/issues/15) ready.

This PR also fixes a bug in the `matrix_cl` copy_assignment that was introduced with the removal of const qualifiers. Currently the buffers size was not modified on copy assignment.

## Tests

The following tests are added in opencl/prim: 
- non-inplace cholesky (exceptions tests, zero sized matrix, ...)
- mdivide_left_tri_low 
- mdivide_right_tri_low

## Side Effects

/

## Checklist

- [x] Math issue #1352 

- [x] Copyright holder: Rok Češnovar (Univ. of Ljubljana)

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/opencl/prim/mdivide_left_tri_low.hpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * @throws std::domain_error if A is not square or the rows of b don't
+ * match the size of A.
+ */
+template <typename T1, typename T2,
+          typename = enable_if_all_floating_point<T1, T2>>
+inline matrix_cl<return_type_t<T1, T2>> mdivide_left_tri_low(
+    const matrix_cl<T1>& A, const matrix_cl<T2>& b) {
+  check_square("mdivide_left_tri_low", "A", A);
+  check_multiplicable("mdivide_left_tri_low", "A", A, "b", b);
+  return tri_inverse<matrix_cl_view::Lower>(A) * b;
```
Comment (rok-cesnovar) [inline stan/math/opencl/prim/mdivide_left_tri_low.hpp]: In order to force the Lower I had to extend the tri_inverse with a template.

The other option would be to force  a change of A before calling tri_inverse but that changes the view for A globally, which is bad. 

The third option would be to have the "forced" view as an argument to tri_inverse. I am also fine with that.

File: stan/math/opencl/prim/multiply.hpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
+#ifndef STAN_MATH_OPENCL_PRIM_MULTIPLY_HPP
+#define STAN_MATH_OPENCL_PRIM_MULTIPLY_HPP
```
Comment (rok-cesnovar) [inline stan/math/opencl/prim/multiply.hpp]: This code was just moved from /opencl/multiply.hpp

File: test/unit/math/opencl/prim/cholesky_decompose_test.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#include <stan/math/opencl/opencl.hpp>
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <vector>
+
+#define EXPECT_MATRIX_NEAR(A, B, DELTA) \
+  for (int i = 0; i < A.size(); i++)    \
+    EXPECT_NEAR(A(i), B(i), DELTA);
+
+TEST(MathMatrix, cholesky_decompose_cl_expections) {
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/prim/cholesky_decompose_test.cpp]: These are the same tests as for the in-place cholesky, just on the non-inplace one.

File: test/unit/math/opencl/prim/mdivide_left_tri_low_test.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
+#ifdef STAN_OPENCL
+#include <stan/math/prim/mat.hpp>
+#include <stan/math/opencl/opencl.hpp>
+#include <test/unit/math/expect_near_rel.hpp>
+#include <boost/random/mersenne_twister.hpp>
+#include <gtest/gtest.h>
+#include <algorithm>
+
+#define EXPECT_MATRIX_NEAR(A, B, DELTA) \
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/prim/mdivide_left_tri_low_test.cpp]: These tests and the mdivide_right_tri_low follow the same idea as the OpenCL tests we have for math/prim/mdivide_left_tri

File: stan/math/opencl/prim/cholesky_decompose.hpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    return res;
+  }
+  opencl::cholesky_decompose(res);
+  // check_pos_definite on matrix_cl is check_nan + check_diagonal_zeros
+  check_nan("cholesky_decompose (OpenCL)", "A", res);
+  check_diagonal_zeros("cholesky_decompose (OpenCL)", "A", res);
+  // TODO(Rok): replace the below two lines with
+  // res.zeros_stric_tri<matrix_cl_view::Lower>();
+  // once that is merged
+  res.view(matrix_cl_view::Lower);
```
Comment (t4c1) [inline stan/math/opencl/prim/cholesky_decompose.hpp]: Should't the implementation of the cholesky already set this view?

File: stan/math/opencl/prim/mdivide_left_tri_low.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/opencl/prim/mdivide_left_tri_low.hpp]: I like the option you chose.

File: stan/math/opencl/prim/cholesky_decompose.hpp (hunk 5 above)
Comment (rok-cesnovar) [reply stan/math/opencl/prim/cholesky_decompose.hpp]: Yes it does. Good call, this is redundant.

Comment (rok-cesnovar) [reply stan/math/opencl/prim/cholesky_decompose.hpp]: Done.

**Comment by rok-cesnovar** [comment]
This is ready for a re-review. But not a critical thing for 2.21.  @SteveBronder you are probably busy these days, take a look whenever you have time. Dont wont to merge this one without your input.

**Comment by SteveBronder** [comment]
I'll try to take a look tmrw if not then Monday

**Comment by rok-cesnovar** [comment]
I would do that once we add compiler support to use them directly. I will work on that in the coming weeks. Then we can finnaly remove the ifdefed code from CPU /prim files. 

And then I would be fine moving /opencl/prim files to /prim/opencl.

**Pull request merged**